drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -90,76 +99,113 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
+static int bq274xx_read_block(const struct device *dev,
+			      uint8_t subclass,
+			      uint8_t *block, uint8_t num_bytes)
 {
+	const struct bq274xx_config *const config = dev->config;
 	int ret;
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_DEVICE_TYPE);
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_CLASS, subclass);
 	if (ret < 0) {
-		LOG_ERR("Unable to write control register");
+		LOG_ERR("Failed to update state subclass");
 		return -EIO;
 	}
 
-	ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_CONTROL_LOW, val);
+	/* DataBlock(), 0 for the first 32 bytes. */
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_BLOCK, 0x00);
 	if (ret < 0) {
-		LOG_ERR("Unable to read register");
+		LOG_ERR("Failed to update block offset");
+		return -EIO;
+	}
+
+	k_sleep(BQ274XX_SUBCLASS_DELAY);
+
+	ret = i2c_burst_read_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
+	if (ret < 0) {
+		LOG_ERR("Unable to read block data");
 		return -EIO;
 	}
 
 	return 0;
 }
 
-static int bq274xx_gauge_configure(const struct device *dev)
+static int bq274xx_write_block(const struct device *dev,
+			       uint8_t *block, uint8_t num_bytes)
 {
 	const struct bq274xx_config *const config = dev->config;
-	struct bq274xx_data *data = dev->data;
-	const struct bq274xx_regs *regs = data->regs;
+	uint8_t checksum = 0;
 	int ret;
-	uint8_t tmp_checksum, checksum_old, checksum_new;
-	uint16_t flags, designenergy_mwh, taperrate, reg_val;
-	uint8_t block[BQ27XXX_DM_SZ];
-	uint8_t try;
 
-	designenergy_mwh = (uint32_t)config->design_capacity * 37 / 10; /* x3.7 */
-	taperrate = config->design_capacity * 10 / config->taper_current;
-
-	/* Unseal the battery control register */
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_A);
+	ret = i2c_burst_write_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
 	if (ret < 0) {
-		LOG_ERR("Unable to unseal the battery");
+		LOG_ERR("Unable to read block data");
 		return -EIO;
 	}
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_B);
+	for (uint8_t i = 0; i < num_bytes; i++) {
+		checksum += block[i];
+	}
+	checksum = 0xff - checksum;
+
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_CHECKSUM, checksum);
 	if (ret < 0) {
-		LOG_ERR("Unable to unseal the battery");
+		LOG_ERR("Failed to update block checksum");
 		return -EIO;
 	}
 
-	/* Send CFG_UPDATE */
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_SET_CFGUPDATE);
+	k_sleep(BQ274XX_SUBCLASS_DELAY);
+
+	return 0;
+}
+
+static void bq274xx_update_block(uint8_t *block,
+				 uint8_t offset, uint16_t val,
+				 bool *block_modified)
+{
+	uint16_t old_val;
+
+	old_val = sys_get_be16(&block[offset]);
+
+	LOG_DBG("update block: off=%d old=%d new=%d\n", offset, old_val, val);
+
+	if (val == old_val) {
+		return;
+	}
+
+	sys_put_be16(val, &block[offset]);
+
+	*block_modified = true;
+}
+
+static int bq274xx_change_mode(const struct device *dev, uint8_t mode)
```

Comment:
A comment to explain why the read flags command is retried 100 times would be good if possible.

Also noticing if it fails it waits up to `100 * 50ms = 5s` which seems very extreme?

Might be out of the scope of this PR but it might be good to make it easy to observe the number of tries required:
```
	for (try = 0; try < MAX_TRIES; try++) {
		ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_FLAGS, &flags);
		if (ret < 0) {
			LOG_ERR("Unable to read flags");
			return -EIO;
		}

		if (flags & BQ27XXX_FLAG_CFGUP != 0x0000) {
			LOG_DBG("success, try %u", try);
			break;
		} else {
			k_sleep(BQ274XX_CFGUP_DELAY);
		}
	}

	if (try >= MAX_TRIES) {
		LOG_ERR("Config mode change timeout");
		return -EIO;
	}
```

Reply:
I had a quick look in https://www.ti.com/lit/gpn/BQ27427 to try to understand this better but could not find a ref for BQ274XX_CFGUP_DELAY

Reply:
Yeah, lot of tribal knowledge in this driver. Linux uses 100x25ms, good enough for them, good enough for us. Changed and used your tweaked code as well, I like it better than the do-while.

Reply:
In TI document SLUUAC9A it says the bit 4 flag change may take up to 1s, see steps 3 and 13:

<img width="629" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/131832/3f2f12ba-371a-4586-ae03-467950d9c77b">

So if BQ274XX_CFGUP_DELAY is 25ms we could reduce MAX_TRIES to 41.

Reply:
That's fine we don't need to snipe the value there, it's a fault condition anyway, aligning with the Linux driver feels safe.